server/src/main/java/org/apache/cloudstack/jsinterpreter/JsInterpreterHelper.java

```diff
@@ -0,0 +1,235 @@
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class JsInterpreterHelper {
+    private final Logger logger = LogManager.getLogger(getClass());
+
+    private int callExpressions;
+
+    private StringBuilder variable;
+
+    private Set<String> variables;
+
+    /**
+     * Returns all variables from the given script.
+     *
+     * @param script the script to extract the variables.
+     * @return A {@link Set<String>} containing all variables in the script.
+     */
+    public Set<String> getScriptVariables(String script) {
+        String parseTree = getScriptAsJsonTree(script);
+        ObjectMapper mapper = new ObjectMapper();
+        JsonNode jsonNode = null;
+        variables = new HashSet<>();
+        variable = new StringBuilder();
+
+        try {
+            jsonNode = mapper.readTree(parseTree);
+        } catch (JsonProcessingException e) {
+            logger.error("Unable to create the script JSON tree due to: [{}].", e.getMessage(), e);
+        }
+
+        logger.trace("Searching script variables from [{}].", script);
+        iterateOverJsonTree(jsonNode.fields());
+
+        if (StringUtils.isNotBlank(variable.toString())) {
+            logger.trace("Adding variable [{}] into the variables set.", variable);
+            removeCallFunctionsFromVariable();
+            variables.add(variable.toString());
+        }
+
+        logger.trace("Found the following variables from the given script: [{}]", variables);
+        return variables;
+    }
+
+    private String getScriptAsJsonTree(String script) {
+        logger.trace("Creating JSON Tree for script [{}].", script);
+        Options options = new Options("nashorn");
+        options.set("anon.functions", true);
+        options.set("parse.only", true);
+        options.set("scripting", true);
+
+        ErrorManager errors = new ErrorManager();
+        Context context = new Context(options, errors, Thread.currentThread().getContextClassLoader());
+        Context.setGlobal(context.createGlobal());
+
+        return ScriptUtils.parse(script, "nashorn", false);
+    }
+
+    protected void iterateOverJsonTree(Iterator<Map.Entry<String, JsonNode>> iterator) {
+        while (iterator.hasNext()) {
+            iterateOverJsonTree(iterator.next());
+        }
+    }
+
+    protected void iterateOverJsonTree(Map.Entry<String, JsonNode> fields) {
+        JsonNode node = null;
+
+        if (fields.getValue().isArray()) {
+            iterateOverArrayNodes(fields);
+        } else {
+            node = fields.getValue();
+        }
+
+        String fieldName = searchIntoObjectNodes(node);
+
+        if (fieldName == null) {
+            String key = fields.getKey();
+            if ("type".equals(key) && "CallExpression".equals(node.textValue())) {
+                callExpressions++;
+            }
+
+            if ("name".equals(key) || "property".equals(key)) {
```

Comment:
You could extract the Strings constants to variable, as they are used in other methods of this class.